src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs

```diff
@@ -45,5 +50,402 @@ protected override void InsertItem(int index, IDeltaSetItem item)
+                    //            ]
+                    //        }
+                    //    ]
+                    //}
+
+                    // If we have a request payload above and we are handling the changed values in Orders,
+                    // The apiHandlerOfT is OrdersAPIHandler with parent as Customers(1).
+                    // The object with "@odata.id" will have OrdersAPIHandler with parent as Customers(3).
+                    // The object with id 2000 will have OrdersAPIHandler similar to apiHandlerOfT.
+                    // The codebelow ensures we use the correct handler.
+                    // NOTE: Handler when we have odata.id will always be different than the parent handler.
+                    // However this is true if handlers are applied correctly.
+
+                    ODataAPIHandler<TStructuralType> odataPathApiHandler = apiHandlerFactory.GetHandler(changedObj.ODataPath) as ODataAPIHandler<TStructuralType>;
+
+                    if (odataPathApiHandler != null && changedObj.ODataPath.Any())
+                    {
+                        // Using a 3rd party library to compare objects
+                        // https://github.com/GregFinzer/Compare-Net-Objects/wiki/Getting-Started
+                        // TODO:
+                        // a) Confirm to perf regressions are caused byt his library.
+                        // b) Investigate if there is a better library.
+
+                        KellermanSoftware.CompareNetObjects.CompareLogic compare = new KellermanSoftware.CompareNetObjects.CompareLogic();
+                        compare.Config.CaseSensitive = false;
+                        compare.Config.ComparePrivateProperties = true;
+                        compare.Config.ComparePrivateFields = true;
+
+                        KellermanSoftware.CompareNetObjects.ComparisonResult result = compare.Compare(handler, odataPathApiHandler);
+
+                        if (!result.AreEqual)
+                        {
+                            handler = odataPathApiHandler;
+                            hasODataId = true;
+                        }
+                    }
+                }
+
+                DataModificationOperationKind operation = DataModificationOperationKind.Update;
+
+                //Get filtered item based on keys
+                TStructuralType original = null;
+                string errorMessage = string.Empty;
+                string getErrorMessage = string.Empty;
+
+                try
+                {
+                    Dictionary<string, object> keyValues = new Dictionary<string, object>();
+                    keyValues = changedObj.ODataPath.GetKeys();
+
+                    bool containsKeyValue = false;
+
+                    foreach (string key in _keys)
+                    {
+                        keyValues.TryGetValue(key, out object value);
+
+                        if (!string.IsNullOrEmpty(value.ToString()))
```

Comment:
if value returned here is null, doing value.ToString() will throw a nullReferenceException